tempesta_fw/http_sticky.c

```diff
@@ -78,67 +78,87 @@ tfw_http_sticky_send_302(TfwHttpMsg *hm)
 	return 0;
 }
 
-/*
- * Find a specific non-special header field in an HTTP message.
- *
- * This function assumes that the header field name is stored
- * in TfwStr{} after an HTTP message is parsed.
- */
-static TfwStr *
-tfw_http_field_raw(TfwHttpMsg *hm, const char *field_name, size_t len)
+static bool
+search_cookie(TfwPool *pool, const TfwStr *cookie, TfwStr *val)
 {
-	int i;
-
-	for (i = TFW_HTTP_HDR_RAW; i < hm->h_tbl->size; i++) {
-		TfwStr *hdr_field = &hm->h_tbl->tbl[i];
-		if (tfw_str_eq_cstr(hdr_field, field_name, len,
-				    TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI))
-			return hdr_field;
-	}
-
-	return NULL;
-}
-
-static int
-tfw_http_field_value(TfwHttpMsg *hm, const TfwStr *field_name, TfwStr *value)
-{
-	char *buf, *ptr;
-	size_t len;
-	TfwStr *hdr_field;
-
-	hdr_field = tfw_http_field_raw(hm, field_name->ptr, field_name->len);
-	if (hdr_field == NULL) {
-		return 0;
-	}
-
-	if (TFW_STR_PLAIN(hdr_field)) {
-		buf = (char *)hdr_field->ptr + field_name->len;
-		len = hdr_field->len - field_name->len;
-
-		while (len > 0 && isspace(*buf)) {
-			buf++;
-			len--;
-		}
-		while (len > 0 && isspace(*(buf + len - 1)))
-			len--;
+	const char *const cstr = tfw_cfg_sticky.name.ptr;
+	const unsigned int clen = tfw_cfg_sticky.name.len + 1;
+
+	const TfwStr *chunk, *next;
+	TfwStr *s = NULL;
+	unsigned int n = TFW_STR_CHUNKN(cookie) + 1;
+	TfwStr tmp;
+	enum {
+		StateName,
+		StateSearchVal,
+		StateVal,
+		StateValChunks,
+	} state = StateName;
+
+	BUG_ON(!TFW_STR_PLAIN(&tfw_cfg_sticky.name));
+
+	TFW_STR_INIT(&tmp);
+	tmp.flags = __TFW_STR_COMPOUND;
+
+	TFW_STR_FOR_EACH_CHUNK(chunk, cookie, {
+		--n;
+		switch (state) {
+
+		case StateName:
+			if (chunk->flags & TFW_STR_NAME) {
+				/*
+				 * Create temporary compound string, starting
+				 * with this chunk.
+				 * We do not use it's overall length now,
+				 * so do not set it.
+				 */
+				tmp.ptr = (void*)chunk;
+				__TFW_STR_CHUNKN_SET(&tmp, n);
+				if (tfw_str_eq_cstr(&tmp, cstr, clen,
```

Comment:
There is no sense to call relatively expensive ```tfw_str_eq_cstr()```: ```cstr``` is plain C-string and ```chunk``` is also a plain string (we don't support chunked chunks).